File: backend/app/ocr/pdf.py
"""Digital-PDF text extraction.

Most real permits are digital PDFs, so this is fast and lossless. Uses pypdf
first, then falls back to pdfplumber. Scanned PDFs with no text layer return
empty text + low confidence, which the pipeline treats as "needs review".
"""
from __future__ import annotations

import io

from ..logger import logger
from .types import OcrPageResult, OcrResult


def _clean(text: str) -> str:
    return " ".join((text or "").split()).strip()


def extract_pdf_text(data: bytes) -> OcrResult:
    pages: list[OcrPageResult] = []

    # Primary: pypdf
    try:
        from pypdf import PdfReader

        reader = PdfReader(io.BytesIO(data))
        for i, page in enumerate(reader.pages):
            text = _clean(page.extract_text() or "")
            pages.append(OcrPageResult(page=i + 1, text=text, confidence=0.97 if text else 0.0))
    except Exception as exc:  # noqa: BLE001
        logger.warn("ocr.pypdf_failed", error=str(exc))
        pages = []

    full_text = "\n".join(p.text for p in pages).strip()

    # Fallback: pdfplumber (better on some layouts)
    if not full_text:
        try:
            import pdfplumber

            with pdfplumber.open(io.BytesIO(data)) as pdf:
                pages = []
                for i, page in enumerate(pdf.pages):
                    text = _clean(page.extract_text() or "")
                    pages.append(OcrPageResult(page=i + 1, text=text, confidence=0.95 if text else 0.0))
            full_text = "\n".join(p.text for p in pages).strip()
        except Exception as exc:  # noqa: BLE001
            logger.warn("ocr.pdfplumber_failed", error=str(exc))

    if full_text:
        return OcrResult(
            text=full_text,
            confidence=sum(p.confidence for p in pages) / max(len(pages), 1),
            page_count=len(pages),
            pages=pages,
            engine="pdf-text",
        )
    return OcrResult(
        text="",
        confidence=0.0,
        page_count=len(pages),
        pages=pages,
        engine="empty",
        note="No embedded text layer (likely a scanned PDF). Use an image of the page for OCR.",
    )

File: lib/ocr/pdf.ts
import type { OcrResult, OcrPageResult } from "./types";
import { logger } from "../logger";

// Node 21 lacks Promise.withResolvers, which the bundled pdfjs uses. Polyfill it
// once so PDF parsing works on Node < 22 too.
if (typeof (Promise as any).withResolvers !== "function") {
  (Promise as any).withResolvers = function withResolvers<T>() {
    let resolve!: (v: T | PromiseLike<T>) => void;
    let reject!: (e?: unknown) => void;
    const promise = new Promise<T>((res, rej) => {
      resolve = res;
      reject = rej;
    });
    return { promise, resolve, reject };
  };
}

// Extracts the embedded text layer from a digital PDF using unpdf (a Node/
// serverless-friendly pdfjs build — no worker file or DOM required). Most real
// permits are digital PDFs, so this is fast and lossless. Scanned PDFs with no
// text layer return empty text + low confidence, which the pipeline treats as
// "needs review" rather than failing.
export async function extractPdfText(buffer: Buffer): Promise<OcrResult> {
  try {
    const { getDocumentProxy, extractText } = await import("unpdf");
    const pdf = await getDocumentProxy(new Uint8Array(buffer));
    const { text, totalPages } = await extractText(pdf, { mergePages: false });

    const perPage: string[] = Array.isArray(text) ? text : [String(text ?? "")];
    const pages: OcrPageResult[] = perPage.map((t, i) => {
      const clean = (t ?? "").replace(/\s+/g, " ").trim();
      return { page: i + 1, text: clean, confidence: clean.length > 0 ? 0.97 : 0 };
    });

    const fullText = pages.map((p) => p.text).join("\n").trim();
    const hasText = fullText.length > 0;
    return {
      text: fullText,
      confidence: hasText ? 0.97 : 0,
      pageCount: totalPages ?? pages.length,
      pages,
      engine: hasText ? "pdf-text" : "empty",
      note: hasText
        ? undefined
        : "No embedded text layer (likely a scanned PDF). Image OCR of PDF pages is not enabled in this build.",
    };
  } catch (err) {
    logger.warn("ocr.pdf_failed", { error: (err as Error).message });
    return { text: "", confidence: 0, pageCount: 0, pages: [], engine: "empty", note: "PDF parsing failed" };
  }
}
